src/libprojectM/Audio/MilkdropFFT.hpp

```diff
@@ -0,0 +1,159 @@
+     * 11,025 Hz range.
+     *
+     * If you want to get 3 bands, try it this way:
+     *   a) 11,025 / 200 = 55.125
+     *   b) to get the number of octaves between 200 and 11,025 Hz, solve for n:
+     *          2^n = 55.125
+     *          n = log 55.125 / log 2
+     *          n = 5.785
+     *   c) so each band should represent 5.785/3 = 1.928 octaves; the ranges are:
+     *          1) 200 - 200*2^1.928                    or  200  - 761   Hz
+     *          2) 200*2^1.928 - 200*2^(1.928*2)        or  761  - 2897  Hz
+     *          3) 200*2^(1.928*2) - 200*2^(1.928*3)    or  2897 - 11025 Hz
+     *
+     * A simple sine-wave-based envelope is convolved with the waveform
+     * data before doing the FFT, to emeliorate the bad frequency response
+     * of a square (i.e. nonexistent) filter.
+     *
+     * You might want to slightly damp (blur) the input if your signal isn't
+     * of a very high quality, to reduce high-frequency noise that would
+     * otherwise show up in the output.
+     * @param waveformData The waveform data to convert. Must contain at least the number of elements passed as samplesIn in Init().
+     * @param spectralData The resulting frequency data. Vector will be resized to samplesOut elements as passed to Init().
+     *                     If the conversion failed, e.g. not initialized or too few input samples, the result vector will be empty.
+     */
+    void TimeToFrequencyDomain(const std::vector<float>& waveformData, std::vector<float>& spectralData);
+
+    /**
+     * @brief Returns the number of frequency samples calculated.
+     * This is twice the value of samplesOut passed to Init().
+     * @return The number of frequency samples calculated.
+     */
+    auto NumFrequencies() const -> size_t
+    {
+        return m_numFrequencies;
+    };
+
+private:
+    /**
+     * @brief Initializes the equalizer envelope table.
+     *
+     * This pre-computation is for multiplying the waveform sample
+     * by a bell-curve-shaped envelope, so we don't see the ugly
+     * frequency response (oscillations) of a square filter.
+     *
+     * - a power of 1.0 will compute the FFT with exactly one convolution.
+     * - a power of 2.0 is like doing it twice; the resulting frequency
+     *   output will be smoothed out and the peaks will be "fatter".
+     * - a power of 0.5 is closer to not using an envelope, and you start
+     *   to get the frequency response of the square filter as 'power'
+     *   approaches zero; the peaks get tighter and more precise, but
+     *   you also see small oscillations around their bases.
+     * @param power Number of convolutions in the envelope.
+     */
+    void InitEnvelopeTable(float power);
+
+    /**
+     * @brief Calculates equalization factors for each frequency domain.
+     *
+     * @param equalize If false, all factors will be set to 1, otherwise will calculate a frequency-based multiplier.
```

Comment:
what's the purpose of the equalizer?

Reply:
From what I understand of the code, it adds a log-based scaling curve over the frequencies. I don't know the reasoning behind this, so I just kept the code as-is to match Milkdrop's spectrum analyzer results.

The equalizer curve which is applied to the frequency bands looks like this:
![image](https://github.com/projectM-visualizer/projectm/assets/5038276/0e501221-a07b-4d7a-938d-0792c2e6ef55)

Reply:
The envelope table, which is also used for scaling the input waveform samples (basically using more of the waveform's center than the edges), looks like this:
![image](https://github.com/projectM-visualizer/projectm/assets/5038276/0713d129-5ce3-4cef-8056-a8762d7cd91b)